### PR TITLE
feat(interpreter-cid,interpreter-data)!: Support for mulitple hash types in CID verification

### PR DESCRIPTION
**Pull request by monoid**
Support multiple hashing algorithms for CID verification.  Will make changing a CID hash algorithm a non-breaking change.

### CODE REVIEW

File: crates/air-lib/interpreter-cid/src/verify.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            &json!(1),
+        )
+        .unwrap();
+        verify_value(
+            &CID::new("z3v8BBK3kqxb39bomB9bJQ22a734aidv5C7QmjdfKiePgVjdQUQ"),
+            &json!({"key": 42}),
+        )
+        .unwrap();
+    }
+}
```
Comment (monoid) [inline crates/air-lib/interpreter-cid/src/verify.rs:140]: TODO: add negative tests.

File: crates/air-lib/interpreter-cid/src/lib.rs [hunk 2]
```diff
@@ -102,60 +120,67 @@ pub fn json_data_cid<Val: ?Sized>(data: &[u8]) -> CID<Val> {
...
-}
-
-impl std::error::Error for CidCalculationError {
-    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-        Some(&self.0)
-    }
+#[derive(Debug, ThisError)]
+pub enum CidCalculationError {
+    #[error(transparent)]
+    Json(#[from] serde_json::Error),
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-cid/src/lib.rs:131]: Let's clarify it a bit, e.g., `InvalidJson`

File: crates/air-lib/interpreter-cid/src/verify.rs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+use serde::Serialize;
+use thiserror::Error as ThisError;
+
+use std::convert::TryInto;
+use std::rc::Rc;
+
+#[derive(ThisError, Debug)]
+pub enum CidVerificationError {
+    #[error("Value mismatch in the {type_name:?} store for CID {cid_repr:?}")]
+    MismatchError {
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-cid/src/verify.rs:28]: I think we can avoid suffix `error` in error variants, we can name this one as `ValueMismatch`

File: crates/air-lib/interpreter-cid/src/verify.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        // nb: type_name is std::any::type_name() result that may be inconsistent between the Rust compiler versions
+        type_name: &'static str,
+        cid_repr: Rc<CidRef>,
+    },
+
+    #[error("JSON error: {0}")]
+    JsonError(#[from] serde_json::Error),
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-cid/src/verify.rs:35]: ditto with `InvalidJson`

File: crates/air-lib/interpreter-cid/src/lib.rs [hunk 5]
```diff
@@ -102,60 +120,67 @@ pub fn json_data_cid<Val: ?Sized>(data: &[u8]) -> CID<Val> {
...
 
     let cid = Cid::new_v1(JSON_CODEC, digest);
     Ok(CID::new(cid.to_string()))
 }
+
+pub(crate) fn value_json_hash<D: digest::Digest + std::io::Write, Val: Serialize + ?Sized>(
+    value: &Val,
+) -> Result<Vec<u8>, serde_json::Error> {
+    let mut hasher = D::new();
+    serde_json::to_writer(BufWriter::with_capacity(8 * 1024, &mut hasher), value)?;
```
Comment (mikevoronov) [inline crates/air-lib/interpreter-cid/src/lib.rs:157]: let's introduce a new local const for `8*1024`.

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**